### DISCUSSION TITLE
Integrating TypeScript into expo start

### DISCUSSION THREAD

**Discussion by brentvatne** [expo/expo-cli#3261] [Ideas]
https://twitter.com/natebosscher/status/1365023487451103239

> it would be nice if `expo start` also ran type checking if your project was configured to use typescript.
>
> I didn't realize that I had to run `tsc` separately to typecheck the entire project. My ide was type checking individual files, so I figured it was just working

**Comment by ntbosscher** [comment]
Also from our discussion:
- Would be nice if `expo build` and friends would fail when type checking fails
   - Should include an escape hatch, something like `--ignore-typescript-errors`

There might be something to learn from how [create-react-app](https://github.com/facebook/create-react-app) does things with typescript

**Reply by brentvatne** [reply]
i'm not totally convinced that this should be the default behaviour - i personally prefer to be able to quickly hack something together with type errors at first and then go back and fix those. i think it interrupts the development flow to have to care about this all the time. i'd be open to contributions that make it possible to extend expo-cli to integrate this type of behavior though! any interest in taking a stab at that?

**Reply by ntbosscher** [reply]
To be clear, I'm only suggesting type-checking on projects that were explicitly configured to use typescript. In my view the primary reason you use typescript is to get the type-checking. So if you configure your project to use TS, I think it would only make sense that you do the type checking. 

Of course some people won't want the type-checking, but then I think you have a flag for that. But, I think those people are outliers. Or am I missing a reason for using typescript other than type-checking?

I'm happy to contribute where I can. I just want to make sure there's some consensus around this idea before I make the effort.

**Reply by brentvatne** [reply]
i think the key factor here is type checking that is blocking on failure or not. i think it's a feature and not a bug that you can run your app when type checking fails, and you should have to opt-in to the blocking behavior imo.

**Reply by brentvatne** [reply]
it's still early in this twitter poll but people generally seem to agree that blocking on typechecking in development is not desirable: https://twitter.com/notbrent/status/1365363025339768834

<img width="602" alt="Screen Shot 2021-02-26 at 12 32 38 PM" src="https://user-images.githubusercontent.com/90494/109351795-b7a7fa80-782e-11eb-9b6a-823e26b1ba60.png">

**Reply by ntbosscher** [reply]
What if it was: "If type-checking fails print error in the console but don't block loading"

**Reply by brentvatne** [reply]
i think people would basically all agree with that :)

**Reply by ntbosscher** [reply]
Ok, I'm good with that. 

To be clear by console, I mean the spot where you ran the `expo build` / `expo start` command and the web console that pops up.

Anything we need to consider with that? Should we have a way to disable that functionality? What log level should ts errors be printed at?

**Reply by brentvatne** [reply]
what would be interesting for this feature is if it were built in such a way that the same infrastructure could be used to support other compile-to-js languages like [reason](https://reasonml.github.io/). that said, typescript is the first-class supported language in the expo ecosystem, and so once we have that built we could enable typechecking in the expo server by default - but without blocking loading the application.

**Reply by brentvatne** [reply]
`start`, `publish` (which is run by `build`), and `export` could be good places to do this. it should definitely be possible to disable it, and we probably don't want to add any additional configuration to app.json - an env var could be a good approach.

**Reply by brentvatne** [reply]
another consideration - i think we would want to try this as something developers can opt-in to for a bit and see how that works before committing to making it the default for typescript apps
